R/tag_env.R

```diff
@@ -0,0 +1,1307 @@
+# * It is easy to alter the current object in place
+# * Difficult to create a new search path while altering in place
+#
+# Disadvantages of standard R objects recursion
+# * Asking for a grandparent element is difficult.
+#   * Altering a grandparent element and having the change stay is impossible
+# * Searching would need to be done at each stage, every time
+
+# ## Using environments elements
+#
+# Advantages
+# * Fast to convert to a "linked list graph" of elements
+# * Access to parents
+# * Calculations on siblings can now be done, even after alterations have been completed
+# * Once a `find(".x")` has been completed, a set of element environment pointers can be stored.
+#   * This makes followup alterations have the minimal O(k) complexity (where k is _found_ elements), not O(n) + O(k) graph search + reconstruction and k _found_ element alterations
+#
+# Disadvantages
+# * MUST be careful not alter the environment object before converting back to a list. (Ex: Do not remove the element environment's children)
+# * The item returned is a set of environments that will alter in place. We will need to be careful about documenting and/or safeguarding this
+
+
+# ## Final design choice:
+# Use environment elements
+# * Being able to search and have a list of eles to immediately look at and alter in place is AMAZING!
+# * Being able to ask for a grandparent (or obj$parent$parent) and be able to alter it in place is AMAZING! This has a strongly influenced by jquery.
+
+# ----------
+
+# # Current design decisions
+# * Tag graphs or tag environments can NOT be used in UI. These objects MUST be converted back to standard tag objects.
+# * tagFunctions will not be altered in place
+#   * To alter tagFunction()s, use the `onRender(x)` method to register a method to be called after `as.tags(x)` is called.
+#   * `onRender(x, expr)` will wrap create a tag function that will resolve the tags before running the expr.
+
+
+# Wrap in function so r cmd check is happy
+envir_key_fn <- function(x) { format.default(x) }
+# No need to contantly convert the envir to a key string. The value should be constant given an envir
+envir_key_or_stop <- function(x) {
+  x$env_key %||% stop("`x` is not a tag environment. `x$env_key` is missing")
+}
+
+# use for `has()` functionality
+envir_map <- function() {
+  map <- fastmap()
+  list(
+    keys = function() {
+      map$keys()
+    },
+    as_list = function() {
+      unname(map$as_list())
+    },
+    has = function(envir) {
+      map$has(envir_key_or_stop(envir))
+    },
+    set = function(envir, value = envir) {
+      map$set(envir_key_or_stop(envir), value)
+    },
+    remove = function(envir) {
+      map$remove(envir_key_or_stop(envir))
+    }
+  )
+}
+# use for consistent `as_list()` order
+envir_stack <- function() {
+  stack <- faststack()
+  list(
+    push = stack$push,
+    as_list = stack$as_list,
+    unique_list = function() {
+      unique(stack$as_list())
```

Comment:
The `unique` is potentially expensive, although for most cases that we're dealing with, I think it's not an issue.

For cases where you want to do this, it might make sense to make a data structure that combines a `fastmap` (for fast duplicate checking when items are added) with a `faststack` (for preserving order). But we'll want to do benchmarks before we decide if that's needed. It could be that the overhead of creating both data structures is more costly than the `unique()` call.